### PR TITLE
Fix: bondfunctions test

### PR DESCRIPTION
**Pull request by dpinte**
@tnicolas not sure what I exactly did but this PR contains all your changes + some fixes I made to the test_bondfunctions.py file.

All the tests pass on my Mac build. I'll try to test that on Windows but I don't see what could break there.

This PR is against the master branch on your repo (tnicolas/pyql master)

### CODE REVIEW

File: quantlib/pricingengines/_bondfunctions.pxd [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+cdef extern from 'ql/cashflows/duration.hpp' namespace 'QuantLib':
+    cdef cppclass Duration:
+        Type type
+
+cdef extern from 'ql/pricingengines/bond/bondfunctions.hpp' namespace 'QuantLib':
+    cdef Rate _bf_yield "BondFunctions::yield" (QLBond, Real, DayCounter, int, Frequency, _dt.Date, Real, Size, Rate)
+
+cdef extern from 'ql/pricingengines/bond/bondfunctions.hpp' namespace 'QuantLib':
+
+    cdef cppclass BondFunctions:
```
Comment (dpinte) [inline quantlib/pricingengines/_bondfunctions.pxd:27]: This works but is not correct. You should have all the functions defined as the `_bf_yield` function. That will also simplify the code on the bonfunctions.pyx file where you can have simple python functions and no need for a Cython Extension class.

File: quantlib/pricingengines/bondfunctions.pyx [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+cimport quantlib.pricingengines.bondfunctions
+cimport quantlib.time._date as _dt
+
+#cdef _bonds.Bond* get_bond(Bond bond):
+#    """ Utility function to extract a properly casted Bond pointer out of the
+#    internal _thisptr attribute of the Instrument base class. """
+#
+#    cdef _bonds.Bond* ref = <_bonds.Bond*>bond._thisptr.get()
+cdef extern from 'ql/compounding.hpp' namespace 'QuantLib':
```
Comment (dpinte) [inline quantlib/pricingengines/bondfunctions.pyx:29]: those cdef's are duplicated from the ones in `_bondfunctions.pxd`. Do you really need them?

File: quantlib/pricingengines/bondfunctions.pyx [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+        print "BondFunctions contructor"
+        self._thisptr = new _bf.BondFunctions()
+
+    def __dealloc__(self):
+        if self._thisptr is not NULL:
+            del self._thisptr
+            self._thisptr = NULL
+
+
+    @classmethod
```
Comment (dpinte) [inline quantlib/pricingengines/bondfunctions.pyx:62]: to be removed

File: quantlib/test/test_bondfunctions.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+
+        pyc_zspd_disco=bfs.zSpread(bond,95.0,ts,ActualActual(ISDA),
+        Compounded,Semiannual,settlement_date,1e-6,100,0.5)
+                                                          
+        # tests
+        #self.assertTrue(Date(27, January, 2011), bond.issue_date)
+        #self.assertTrue(Date(31, August, 2020), bond.maturity_date)
+        #self.assertTrue(settings.evaluation_date, bond.valuation_date)
+
+        print settings.evaluation_date
```
Comment (dpinte) [inline quantlib/test/test_bondfunctions.py:176]: ! that code does not compile on Python 3 (see travis-ci failure). You should have:
`from __future__ import print_function` 

and then use only the print function:

```
       print(d)
       print ('Principal: {}'.format(face_amount))
...
```

File: quantlib/test/test_bondfunctions.py [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+        print d
+        print 'Principal: {}'.format(face_amount)
+        print 'Issuing date: {} '.format(bond.issue_date)
+        print 'Maturity: {}'.format(bond.maturity_date)
+        print 'Coupon rate: {:.4%}'.format(coupon_rate)
+        print 'Yield: {:.4%}'.format(bond_yield)
+        print 'z-spread: {:.4%}'.format(zspd)
+        print 'premium  z-spread using pwyc: {:.4%}'.format(pyc_zspd)
+        print 'discount z-spread using pwyc: {:.4%}'.format(pyc_zspd_disco)
+#        print 'Net present value: {:.4f}'.format(bond.net_present_value)
```
Comment (dpinte) [inline quantlib/test/test_bondfunctions.py:186]: commented lines should be removed

File: setup.py [hunk 6]
```diff
@@ -94,12 +93,12 @@ def get_extra_compile_args():
 
 def get_extra_link_args():
     if sys.platform == 'win32':
-        args = ['/subsystem:windows', '/machine:I386']
+        args = ['/subsystem:windows', '/machine:x64']
```
Comment (dpinte) [inline setup.py:96]: we should add an auto-detection of the bitness of the platform in place of having to change it by hand.

Comment (tnicolas) [inline setup.py:96]: agreed!

On Mar 1, 2015, at 2:07 PM, "Didrik Pinte" <notifications@github.com<mailto:notifications@github.com>> wrote:

In setup.pyhttps://github.com/tnicolas/pyql/pull/1#discussion_r25570076:

> @@ -94,12 +93,12 @@ def get_extra_compile_args():
> 
>  def get_extra_link_args():
>      if sys.platform == 'win32':
> -        args = ['/subsystem:windows', '/machine:I386']
> -        args = ['/subsystem:windows', '/machine:x64']

we should add an auto-detection of the bitness of the platform in place of having to change it by hand.

##

**Pull request merged**